python/pyspark/sql/functions.py

```diff
@@ -1826,19 +1826,27 @@ class UserDefinedFunction(object):
     def __init__(self, func, returnType, name=None):
         self.func = func
         self.returnType = returnType
-        self._judf = self._create_judf(name)
-
-    def _create_judf(self, name):
+        self._judf_placeholder = None
+        self._name = name or (
+            func.__name__ if hasattr(func, '__name__')
+            else func.__class__.__name__)
+
+    @property
+    def _judf(self):
+        if self._judf_placeholder is None:
```

Comment:
I stress tested this a bit and I haven't found any abnormalities but I found a small problem with `__call__` on the way.  Fixed now.